File: packages/tiktok_auth_android/android/src/main/kotlin/dev/tiktokauth/android/PendingAuthStore.kt
// Copyright 2026 The tiktok_auth authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package dev.tiktokauth.android

import android.content.Context
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

/**
 * Persists a sign-in across process death.
 *
 * Android may stop the app while the user is in TikTok or the browser. The
 * attempt in progress is stored so the callback can still be matched when the
 * process is restarted, and the finished result is stored until Dart collects
 * it with `takePendingResult()`. Both entries expire after [MAX_AGE_MILLIS]
 * because TikTok authorization codes are short-lived.
 */
internal class PendingAuthStore(context: Context) {
    private val prefs =
        context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    fun saveInProgress(auth: PendingAuth) {
        // commit() rather than apply(): the process may be stopped right after
        // TikTok or the browser comes to the foreground.
        prefs.edit().putString(KEY_IN_PROGRESS, auth.toJson()).commit()
    }

    fun loadInProgress(nowMillis: Long = System.currentTimeMillis()): PendingAuth? {
        val auth = prefs.getString(KEY_IN_PROGRESS, null)?.let(PendingAuth::fromJson)
        if (auth == null || nowMillis - auth.createdAtMillis > MAX_AGE_MILLIS) {
            clearInProgress()
            return null
        }
        return auth
    }

    fun clearInProgress() {
        prefs.edit().remove(KEY_IN_PROGRESS).apply()
    }

    fun saveCompleted(result: PlatformAuthResult, nowMillis: Long = System.currentTimeMillis()) {
        prefs.edit().putString(KEY_COMPLETED, result.toJson(nowMillis)).commit()
    }

    fun takeCompleted(nowMillis: Long = System.currentTimeMillis()): PlatformAuthResult? {
        val json = prefs.getString(KEY_COMPLETED, null) ?: return null
        prefs.edit().remove(KEY_COMPLETED).apply()
        return try {
            val o = JSONObject(json)
            if (nowMillis - o.getLong(COMPLETED_AT) > MAX_AGE_MILLIS) return null
            val scopes = o.getJSONArray(GRANTED_SCOPES)
            PlatformAuthResult(
                errorKind = PlatformErrorKind.ofRaw(o.getInt(ERROR_KIND)) ?: PlatformErrorKind.FAILED,
                usedWebAuth = o.getBoolean(USED_WEB_AUTH),
                grantedScopes = List(scopes.length()) { scopes.getString(it) },
                authCode = o.stringOrNull(AUTH_CODE),
                codeVerifier = o.stringOrNull(CODE_VERIFIER),
                state = o.stringOrNull(STATE),
                expectedState = o.stringOrNull(EXPECTED_STATE),
                redirectUri = o.stringOrNull(REDIRECT_URI),
                nativeCode = o.stringOrNull(NATIVE_CODE),
                errorDescription = o.stringOrNull(ERROR_DESCRIPTION),
            )
        } catch (e: JSONException) {
            null
        }
    }

    private fun PlatformAuthResult.toJson(nowMillis: Long): String =
        JSONObject()
            .put(ERROR_KIND, errorKind.raw)
            .put(USED_WEB_AUTH, usedWebAuth)
            .put(GRANTED_SCOPES, JSONArray(grantedScopes))
            .putOpt(AUTH_CODE, authCode)
            .putOpt(CODE_VERIFIER, codeVerifier)
            .putOpt(STATE, state)
            .putOpt(EXPECTED_STATE, expectedState)
            .putOpt(REDIRECT_URI, redirectUri)
            .putOpt(NATIVE_CODE, nativeCode)
            .putOpt(ERROR_DESCRIPTION, errorDescription)
            .put(COMPLETED_AT, nowMillis)
            .toString()

    private fun JSONObject.stringOrNull(key: String): String? =
        if (has(key) && !isNull(key)) getString(key) else null

    companion object {
        const val MAX_AGE_MILLIS = 15 * 60 * 1000L

        private const val PREFS_NAME = "dev.tiktokauth.android.pending"
        private const val KEY_IN_PROGRESS = "inProgress"
        private const val KEY_COMPLETED = "completed"

        private const val ERROR_KIND = "errorKind"
        private const val USED_WEB_AUTH = "usedWebAuth"
        private const val GRANTED_SCOPES = "grantedScopes"
        private const val AUTH_CODE = "authCode"
        private const val CODE_VERIFIER = "codeVerifier"
        private const val STATE = "state"
        private const val EXPECTED_STATE = "expectedState"
        private const val REDIRECT_URI = "redirectUri"
        private const val NATIVE_CODE = "nativeCode"
        private const val ERROR_DESCRIPTION = "errorDescription"
        private const val COMPLETED_AT = "completedAtMillis"
    }
}

File: packages/tiktok_auth_ios/ios/tiktok_auth_ios/Sources/tiktok_auth_ios/PendingAuthStore.swift
// Copyright 2026 The tiktok_auth authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation

/// The parts of a sign-in attempt needed to finish it after a relaunch.
struct PendingAuthRecord: Codable, Equatable {
  let id: String
  let redirectUri: String
  let expectedState: String
  let codeVerifier: String
  let usedWebAuth: Bool
  let createdAt: Date
}

/// Persists a sign-in across app relaunches in `UserDefaults`.
///
/// Entries expire after `maxAge` because TikTok authorization codes are
/// short-lived.
final class PendingAuthStore {
  static let maxAge: TimeInterval = 15 * 60

  private static let inProgressKey = "dev.tiktokauth.ios.inProgress"
  private static let completedKey = "dev.tiktokauth.ios.completed"

  private let defaults: UserDefaults
  private let now: () -> Date

  init(defaults: UserDefaults = .standard, now: @escaping () -> Date = Date.init) {
    self.defaults = defaults
    self.now = now
  }

  func saveInProgress(_ record: PendingAuthRecord) {
    defaults.set(try? JSONEncoder().encode(record), forKey: Self.inProgressKey)
  }

  func loadInProgress() -> PendingAuthRecord? {
    guard let data = defaults.data(forKey: Self.inProgressKey),
      let record = try? JSONDecoder().decode(PendingAuthRecord.self, from: data)
    else { return nil }
    guard now().timeIntervalSince(record.createdAt) <= Self.maxAge else {
      clearInProgress()
      return nil
    }
    return record
  }

  func clearInProgress() {
    defaults.removeObject(forKey: Self.inProgressKey)
  }

  func saveCompleted(_ result: PlatformAuthResult) {
    let record = CompletedRecord(result: result, completedAt: now())
    defaults.set(try? JSONEncoder().encode(record), forKey: Self.completedKey)
  }

  func takeCompleted() -> PlatformAuthResult? {
    guard let data = defaults.data(forKey: Self.completedKey) else { return nil }
    defaults.removeObject(forKey: Self.completedKey)
    guard let record = try? JSONDecoder().decode(CompletedRecord.self, from: data),
      now().timeIntervalSince(record.completedAt) <= Self.maxAge
    else { return nil }
    return record.result
  }
}

private struct CompletedRecord: Codable {
  let errorKind: Int
  let usedWebAuth: Bool
  let grantedScopes: [String]
  let authCode: String?
  let codeVerifier: String?
  let state: String?
  let expectedState: String?
  let redirectUri: String?
  let nativeCode: String?
  let errorDescription: String?
  let completedAt: Date

  init(result: PlatformAuthResult, completedAt: Date) {
    errorKind = result.errorKind.rawValue
    usedWebAuth = result.usedWebAuth
    grantedScopes = result.grantedScopes
    authCode = result.authCode
    codeVerifier = result.codeVerifier
    state = result.state
    expectedState = result.expectedState
    redirectUri = result.redirectUri
    nativeCode = result.nativeCode
    errorDescription = result.errorDescription
    self.completedAt = completedAt
  }

  var result: PlatformAuthResult {
    PlatformAuthResult(
      errorKind: PlatformErrorKind(rawValue: errorKind) ?? .failed,
      usedWebAuth: usedWebAuth,
      grantedScopes: grantedScopes,
      authCode: authCode,
      codeVerifier: codeVerifier,
      state: state,
      expectedState: expectedState,
      redirectUri: redirectUri,
      nativeCode: nativeCode,
      errorDescription: errorDescription
    )
  }
}
